### PR TITLE
Code review: support teacher accounts as participants

### PR DESCRIPTION
**Pull request by bencodeorg**
The code review feature in Javalab has made some assumptions that teacher accounts are always operating in the "instructor" role. This PR makes a couple of changes to support teacher accounts being used in the "participant" role (ie, as followers within a section).

General idea is that when a teacher account is in a CSA section and has "view as student" selected, the behavior they observe should be identical to the behavior a student account would observe in the same situation.

Changes made here:
- Comments made by teacher accounts are only saved as "from teacher" if the teacher account writing the comment instructs a section that the project owner is in. This change makes it so special formatting of instructor comments only applies to instructors (not all teacher accounts), and that comments from a teacher account operating as a participant are not hidden from fellow participants (we do want to hide comments from the actual instructor from peers).
- When teacher accounts navigate to a peer's project via the dropdown of projects available for review, we append a URL param to maintain the "view as student" when the teacher account navigates to the peer's project.
- Teachers accounts can delete comments on their projects -- the delete option appears in the comment menu if a participant teacher account selects "view as teacher".

## Links

- code review permissions: [doc](https://docs.google.com/document/d/1yUbfW9urFgieFhyLSoqBO35J-uOJwUWFjpSEd0KQxRI/edit#bookmark=id.742hkd8sl6aw)
- jira ticket: [JAVA-234](https://codedotorg.atlassian.net/browse/JAVA-234)

## Testing story

I went through the code review permissions matrix (linked above) and confirmed that a teacher account enrolled in a section (ie, as a participant) could perform each of the required behaviors for a classroom participant.

### CODE REVIEW

File: apps/src/templates/instructions/ReviewTab.jsx [hunk 1]
```diff
@@ -62,10 +63,19 @@ class ReviewTab extends Component {
...
-    currentLocation().origin +
-    currentLocation().pathname +
-    `?${VIEWING_CODE_REVIEW_URL_PARAM}=true`;
+  generateLevelUrlWithCodeReviewParam = () => {
+    let url =
+      currentLocation().origin +
+      currentLocation().pathname +
+      `?${VIEWING_CODE_REVIEW_URL_PARAM}=true`;
+
+    // If teacher account is viewing as student, set up URLs
```
Comment (dmcavoy) [inline apps/src/templates/instructions/ReviewTab.jsx]: nit: participant instead of student

**Comment by dmcavoy** [comment]
I want to check my understanding. Are you saying that a teacher would need to go in to the teacher panel, click viewAs student, and then their comments would be saved as not from the teacher?

**Comment by bencodeorg** [comment]
> I want to check my understanding. Are you saying that a teacher would need to go in to the teacher panel, click viewAs student, and then their comments would be saved as not from the teacher?

@dmcavoy no, any comment from a teacher account that is not an instructor of a section that the project owner is in would be saved as not from the teacher (regardless of the selected viewAs).

Or, maybe a less convoluted description is...only comments from instructors of the project owner are now saved as "from teacher".

That said, there are UI differences that we still put in place via viewAs -- if you're a teacher account looking for the participant (student) experience of code review, you need to select viewAs = Student. Maybe that goes away once we know the context in which a user is viewing a page (participant vs. instructor)?

In any case, in the short term, I can reach out to PL to let them know that we should advise PL participants to select viewAs = Student when they're trying out the feature as a student?

File: apps/src/templates/instructions/ReviewTab.jsx [hunk 2]
```diff
@@ -62,10 +63,19 @@ class ReviewTab extends Component {
...
+    // to persist this setting when they click to view another project.
+    if (this.props.userIsTeacher && !this.props.viewAsTeacher) {
+      url += `&viewAs=Participant`;
```
Comment (sanchitmalhotra126) [inline apps/src/templates/instructions/ReviewTab.jsx:75]: To clarify - the `viewAs=Participant` flag is not necessary for a participant viewing another participant's project? i.e. student viewing another student's in their section?

**Review by molly-moen — APPROVED** [review]

File: dashboard/app/models/code_review_comment.rb [hunk 3]
```diff
@@ -52,8 +52,10 @@ def self.user_can_review_project?(project_owner, potential_reviewer, project_id,
   def compute_is_from_teacher
     # Only should happen if we hard delete the commenter's account
     # via the delete accounts helper.
-    return false if commenter.nil?
+    return false if commenter.nil? || project_owner.nil?
 
-    self.is_from_teacher = commenter.teacher? ? true : false
+    # This is not ideal, as it only checks that the project owner
+    # is a participant in any section instructed by the commenter.
+    self.is_from_teacher = project_owner.teachers.include?(commenter)
```
Comment (sanchitmalhotra126) [inline dashboard/app/models/code_review_comment.rb:59]: Could we check if the commenter is associated with this project owner's code review group instead? Or would that just have the same effect?

**Review by sanchitmalhotra126 — APPROVED** [review]
LGTM! Thanks for clarifying what's going in the description :)

File: apps/src/templates/instructions/ReviewTab.jsx (hunk 2 above)
Comment (bencodeorg) [reply apps/src/templates/instructions/ReviewTab.jsx:75]: Yeah exactly, it's only relevant if you are logged in with a teacher account and see the blue pullout on the right, which changes the URL (and your view of code review, among other things). Here's a teacher account looking at their own work on a Javalab level with Student selected (looks good, looks like expected experience for a participant/student):

image

Here's teacher account looking at a Javalab level with View as Teacher selected (looks weird, assumes that you are viewing a student's work -- eg, puts your name where it expects a students name in "This is the beginning of feedback for benjamin's project):

image

File: dashboard/app/models/code_review_comment.rb (hunk 3 above)
Comment (bencodeorg) [reply dashboard/app/models/code_review_comment.rb:59]: I guess if we assume that comments can only come from a) the project owner, b) ppl in your review group, or c) your instructor/teacher (which is true at the moment), then yeah I think we could say `is_from_teacher = !in_review_group && !from_you`?

Maybe as-is this is slightly less correct, but seems clearer to me and not conditional on any changes we might make to who can review code?

**Pull request merged**